### PR TITLE
Add API to get/update/delete challenge phase allowed email ids

### PR DESCRIPTION
**Pull request by Ram81**
### Description

This PR -

- [x] Adds API to Get/Update/Delete email ids from `allowed_email_ids` field for a challenge phase.

### CODE REVIEW

File: apps/challenges/urls.py [hunk 1]
```diff
@@ -209,6 +209,11 @@
     url(
         r"^pwc_task_dataset/$", views.pwc_task_dataset, name="pwc_task_dataset"
     ),
+    url(
+        r"^challenge/(?P<challenge_pk>[0-9]+)/challenge_phase/(?P<phase_pk>[0-9]+)/allowed_email_ids/$",
```
Comment (RishabhJain2018) [inline apps/challenges/urls.py:213]: `r"^challenge/(?P<challenge_pk>[0-9]+)/challenge_phase/(?P<phase_pk>[0-9]+)/allowed_email_ids/$",
` -> `r"^challenges/(?P<challenge_pk>[0-9]+)/phases/(?P<phase_pk>[0-9]+)/allowed_email_ids/$"`

File: apps/challenges/views.py [hunk 2]
```diff
@@ -3354,3 +3354,103 @@ def pwc_task_dataset(request):
             "error": "You are not authorized to make this request. Please ask EvalAI admin to add you as a staff user for accessing this API."
         }
         return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+
+@api_view(["GET", "PATCH", "DELETE"])
+@throttle_classes([UserRateThrottle])
+@permission_classes(
+    (
+        permissions.IsAuthenticatedOrReadOnly,
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:3363]: Why `ReadOnly`?

File: apps/challenges/views.py [hunk 3]
```diff
@@ -3354,3 +3354,103 @@ def pwc_task_dataset(request):
...
+        else:
+            response_data = {"error": "Field allowed_email_ids is missing."}
+            return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    if request.method == "GET":
+        serializer = ChallengePhaseCreateSerializer(
+            challenge_phase, context={"request": request}
+        )
+        response_data = {
+            "challenge_phase_pk": phase_pk,
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:3404]: We don't need this `phase_pk` here as the challenge host already knows it while calling the API?

File: apps/challenges/views.py [hunk 4]
```diff
@@ -3354,3 +3354,103 @@ def pwc_task_dataset(request):
...
+            context={"challenge": challenge},
+            partial=True,
+        )
+
+        if serializer.is_valid():
+            serializer.save()
+            challenge_phase = get_challenge_phase_model(serializer.instance.pk)
+            serializer = ChallengePhaseSerializer(challenge_phase)
+            response_data = response_data = {
+                "challenge_phase_pk": phase_pk,
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:3424]: Same as above!

File: apps/challenges/views.py [hunk 5]
```diff
@@ -3354,3 +3354,103 @@ def pwc_task_dataset(request):
...
+            return Response(response_data, status=status.HTTP_200_OK)
+        else:
+            return Response(
+                serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
+
+    elif request.method == "DELETE":
+        remove_allowed_email_ids = request.data["allowed_email_ids"]
+        allowed_email_ids = list(
+            set(allowed_email_ids) - set(remove_allowed_email_ids)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:3436]: Please use `.difference` as it is cleaner.

File: apps/challenges/views.py [hunk 6]
```diff
@@ -3354,3 +3354,103 @@ def pwc_task_dataset(request):
...
+            context={"challenge": challenge},
+            partial=True,
+        )
+
+        if serializer.is_valid():
+            serializer.save()
+            challenge_phase = get_challenge_phase_model(serializer.instance.pk)
+            serializer = ChallengePhaseSerializer(challenge_phase)
+            response_data = response_data = {
+                "challenge_phase_pk": phase_pk,
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:3451]: Same as above.

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
@Ram81 Can we please --
1. Add documentation for this API on https://eval.ai/api/docs/
2. Add an issue for writing tests for this API.

File: apps/challenges/views.py (hunk 2 above)
Comment (Ram81) [reply apps/challenges/views.py]: Yeah, it should be `IsAuthenticated` only.

**Review by RishabhJain2018 — APPROVED** [review]
LGTM!
@Ram81 Can you please open an issue for writing tests of this API?

**Pull request merged**